### PR TITLE
WP_Term_Query cache problem

### PR DESCRIPTION
**Pull request by spacedmonkey**
Trac ticket: https://core.trac.wordpress.org/ticket/55837

---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

### CODE REVIEW

File: tests/phpunit/tests/term/getTerms.php [hunk 1]
```diff
@@ -746,6 +746,26 @@ public function test_get_terms_child_of() {
...
+		$parent = self::factory()->category->create();
+		self::factory()->category->create( array( 'parent' => $parent ) );
+
+		$args  = array(
+			'fields'     => 'ids',
+			'child_of'   => $parent,
+			'hide_empty' => false,
+		);
+		$terms = get_terms( 'category', $args );
+		$this->assertCount( 1, $terms, 'Check count on first response' );
```
Comment (costdev) [inline tests/phpunit/tests/term/getTerms.php]:
```suggestion
		$this->assertCount( 1, $terms, 'The first call to get_terms() did not return 1 term' );
```
The `message` parameter should indicate what went wrong.

File: tests/phpunit/tests/term/getTerms.php [hunk 2]
```diff
@@ -746,6 +746,26 @@ public function test_get_terms_child_of() {
...
+
+		$terms2 = get_terms( 'category', $args );
+		$this->assertCount( 1, $terms2, 'Check count on cached response' );
+		$this->assertSame( $terms, $terms2, 'Results are the same after caching' );
```
Comment (costdev) [inline tests/phpunit/tests/term/getTerms.php]:
```suggestion
		$this->assertCount( 1, $terms2, 'The second call to get_terms() did not return 1 term' );
		$this->assertSame( $terms, $terms2, 'Results are not the same after caching' );
```
The `message` parameter should indicate what went wrong.

File: tests/phpunit/tests/term/getTerms.php [hunk 3]
```diff
@@ -2589,6 +2609,75 @@ public function test_pad_counts() {
 		}
 	}
 
+	/**
+	 * @ticket 55837
+	 */
+	public function test_pad_counts_cached() {
```
Comment (costdev) [inline tests/phpunit/tests/term/getTerms.php:2617]: The assertions in this test method need a `message` parameter to indicate which assertion failed and what went wrong.

File: tests/phpunit/tests/term/getTerms.php [hunk 4]
```diff
@@ -746,6 +746,26 @@ public function test_get_terms_child_of() {
 		$this->assertCount( 1, $terms );
 	}
 
+	/**
+	 * @ticket 55837
```
Comment (costdev) [inline tests/phpunit/tests/term/getTerms.php:750]:
```suggestion
	 * @ticket 55837
	 * @covers ::get_terms
```

File: tests/phpunit/tests/term/getTerms.php [hunk 5]
```diff
@@ -2589,6 +2609,75 @@ public function test_pad_counts() {
 		}
 	}
 
+	/**
+	 * @ticket 55837
```
Comment (costdev) [inline tests/phpunit/tests/term/getTerms.php:2614]:
```suggestion
	 * @ticket 55837
	 * @covers ::get_terms
```

**Review by costdev — CHANGES_REQUESTED** [review]
Thanks for the PR Jonny!

This review suggests some changes in the tests.

File: tests/phpunit/tests/term/getTerms.php [hunk 6]
```diff
@@ -746,6 +746,27 @@ public function test_get_terms_child_of() {
...
+			'fields'     => 'ids',
+			'child_of'   => $parent,
+			'hide_empty' => false,
+		);
+		$terms = get_terms( 'category', $args );
+		$this->assertCount( 1, $terms, 'The first call to get_terms() did not return 1 term' );
+
+		$terms2 = get_terms( 'category', $args );
+		$this->assertCount( 1, $terms2, 'The second call to get_terms() did not return 1 term' );
+		$this->assertSame( $terms, $terms2, 'Results are not the same after caching' );
```
Comment (costdev) [inline tests/phpunit/tests/term/getTerms.php]: Should this be `assertSameSets()`, or?

File: tests/phpunit/tests/term/getTerms.php (hunk 3 above)
Comment (spacedmonkey) [reply tests/phpunit/tests/term/getTerms.php:2617]: Done!

Comment (spacedmonkey) [reply tests/phpunit/tests/term/getTerms.php:2617]: Done!

File: src/wp-includes/class-wp-term-query.php [hunk 7]
```diff
@@ -770,7 +770,7 @@ public function get_terms() {
 		// $args can be anything. Only use the args defined in defaults to compute the key.
 		$cache_args = wp_array_slice_assoc( $args, array_keys( $this->query_var_defaults ) );
 
-		unset( $cache_args['pad_counts'], $cache_args['update_term_meta_cache'] );
+		unset( $cache_args['update_term_meta_cache'] );
```
Comment (peterwilsoncc) [inline src/wp-includes/class-wp-term-query.php:773]: It looks like this changes the cache key & format of the cached data. To be safe the global cache should be flushed during update.

Changing the database version will cause that to happen, no update function is required.
https://github.com/WordPress/wordpress-develop/blob/ae9e0bb74f2a1c1e4721799454d7ec2ee3e5827b/src/wp-includes/version.php#L25-L26

https://github.com/WordPress/wordpress-develop/blob/2fcc00815bd5546ed40296242ff9da66693120df/src/wp-admin/includes/upgrade.php#L663-L667

The chances of a collision a really slim.

Comment (spacedmonkey) [inline src/wp-includes/class-wp-term-query.php:773]: Padding counts changes the response, so we have to have a different cache.

Comment (spacedmonkey) [inline src/wp-includes/class-wp-term-query.php:773]: Padding counts changes the response, so we have to have a different cache.

File: src/wp-includes/class-wp-term-query.php [hunk 8]
```diff
@@ -849,34 +852,11 @@ public function get_terms() {
...
-				if ( isset( $_tt_ids[ $term->term_id ] ) ) {
-					continue;
-				}
-
-				$_tt_ids[ $term->term_id ] = 1;
-				$_terms[]                  = $term;
-			}
-
-			$terms = $_terms;
-		}
```
Comment (spacedmonkey) [inline src/wp-includes/class-wp-term-query.php:871]: This code no longer does anything, so should be removed.

Comment (spacedmonkey) [inline src/wp-includes/class-wp-term-query.php:871]: This code no longer does anything, so should be removed.

**Comment by spacedmonkey** [comment]
@peterwilsoncc 

The biggest change here, is that before there were two variables, $term_objects and $terms. These variables basically did the same thing but depending on what params were passed, would result in different values. This PR, ensure that if `pad_counts` or `child_of` are passed, that the correct value is returned. It also makes the code a lot easier to read, by removing two variables. It also improves cache value, by only caching an array of ids, if other properties are not needed.

**Review by peterwilsoncc — COMMENTED** [review]
@spacedmonkey Thanks for the explanation, it really helped.

Just one note below about triggering a cache flush on update. The current revision ID on SVN is `53479` so you can use that as the DB version for now but it is usually updated on commit to the current SVN revision.

File: src/wp-includes/class-wp-term-query.php (hunk 7 above)
Comment (spacedmonkey) [reply src/wp-includes/class-wp-term-query.php:773]: Unsetting this, will change the cache key. Meaning caches will be invalidated. Meaning we should not have to flush the cache.

**Comment by spacedmonkey** [comment]
Committed https://github.com/WordPress/wordpress-develop/commit/6d89ea8f09cf21ef0d1e1acd6e1b30fba1a72301

**Pull request closed without merging**